docker/build_artifacts/sagemaker/serve.py

```diff
@@ -89,30 +92,33 @@ def __init__(self):
             parts = self._sagemaker_port_range.split("-")
             low = int(parts[0])
             hi = int(parts[1])
-            self._tfs_grpc_port = []
-            self._tfs_rest_port = []
             if low + 2 * self._tfs_instance_count > hi:
                 raise ValueError("not enough ports available in SAGEMAKER_SAFE_PORT_RANGE ({})"
                                  .format(self._sagemaker_port_range))
-            self._tfs_grpc_port_range = "{}-{}".format(low,
-                                                       low + 2 * self._tfs_instance_count)
-            self._tfs_rest_port_range = "{}-{}".format(low + 1,
-                                                       low + 2 * self._tfs_instance_count + 1)
-            for i in range(self._tfs_instance_count):
-                self._tfs_grpc_port.append(str(low + 2 * i))
-                self._tfs_rest_port.append(str(low + 2 * i + 1))
-            # set environment variable for python service
-            os.environ["TFS_GRPC_PORT_RANGE"] = self._tfs_grpc_port_range
-            os.environ["TFS_REST_PORT_RANGE"] = self._tfs_rest_port_range
+            # split ports range here for grpc and rest port
+            rest_port = low
+            grpc_port = (low + hi) // 2
+
+            self._tfs_grpc_port = self._get_random_ports(rest_port, grpc_port)
+            self._tfs_rest_port = self._get_random_ports(grpc_port, hi)
+            self._tfs_grpc_port_range = self._concat_ports(self._tfs_grpc_port)
+            self._tfs_rest_port_range = self._concat_ports(self._tfs_rest_port)
         else:
-            # just use the standard default ports
-            self._tfs_grpc_port = ["9000"]
-            self._tfs_rest_port = ["8501"]
-            self._tfs_grpc_port_range = "9000-9000"
-            self._tfs_rest_port_range = "8501-8501"
-            # set environment variable for python service
-            os.environ["TFS_GRPC_PORT_RANGE"] = self._tfs_grpc_port_range
-            os.environ["TFS_REST_PORT_RANGE"] = self._tfs_rest_port_range
+            if self._tfs_instance_count > 1:
+                self._tfs_grpc_port = self._get_random_ports(9000, 9999)
+                self._tfs_rest_port = self._get_random_ports(8500, 9000)
+                self._tfs_grpc_port_range = self._concat_ports(self._tfs_grpc_port)
+                self._tfs_rest_port_range = self._concat_ports(self._tfs_rest_port)
+            else:
+                # just use the standard default ports
+                self._tfs_grpc_port = ["9000"]
+                self._tfs_rest_port = ["8501"]
+                self._tfs_grpc_port_range = "9000"
+                self._tfs_rest_port_range = "8501"
```

Comment:
This still doesn't conform to the definition of a range, but I guess this was convenient in order to avoid additional checks when parsing range string (generic parser should be able to handle singleton ranges like 9000-9000 and deduce it's just one port...).